content/docs/context.md

```diff
@@ -4,220 +4,130 @@ title: Context
 
-## Referencing Context in Lifecycle Methods
+**theme-context.js**
+`embed:context/theme-detailed-theme-context.js`
 
-If `contextTypes` is defined within a component, the following [lifecycle methods](/docs/react-component.html#the-component-lifecycle) will receive an additional parameter, the `context` object:
+**themed-button.js**
+`embed:context/theme-detailed-themed-button.js`
 
-- [`constructor(props, context)`](/docs/react-component.html#constructor)
-- [`componentWillReceiveProps(nextProps, nextContext)`](/docs/react-component.html#componentwillreceiveprops)
-- [`shouldComponentUpdate(nextProps, nextState, nextContext)`](/docs/react-component.html#shouldcomponentupdate)
-- [`componentWillUpdate(nextProps, nextState, nextContext)`](/docs/react-component.html#componentwillupdate)
+**app.js**
+`embed:context/theme-detailed-app.js`
 
-> Note:
->
-> As of React 16, `componentDidUpdate` no longer receives `prevContext`.
+### Consuming Multiple Contexts
 
-## Referencing Context in Stateless Functional Components
+To keep context re-rendering fast, React needs to make each context consumer a separate node in the tree. 
 
-Stateless functional components are also able to reference `context` if `contextTypes` is defined as a property of the function. The following code shows a `Button` component written as a stateless functional component.
+`embed:context/multiple-contexts.js`
 
-```javascript
-import PropTypes from 'prop-types';
+If two or more context values are often used together, you might want to consider creating your own render prop component that provides both.
 
-const Button = ({children}, context) =>
-  <button style={{background: context.color}}>
-    {children}
-  </button>;
+### Accessing Context in Lifecycle Methods
 
-Button.contextTypes = {color: PropTypes.string};
-```
+Accessing values from context in lifecycle methods is a relatively common use case. Instead of adding context to every lifecycle method, you just need to pass it as a prop, and then work with it just like you'd normally work with a prop.
 
-## Updating Context
+`embed:context/lifecycles.js`
 
-Don't do it.
+### Forwarding Refs to Context Consumers
 
-React has an API to update context, but it is fundamentally broken and you should not use it.
+One issue with the render prop API is that refs don't automatically get passed to wrapped elements. To get around this, use `React.forwardRef`:
 
-The `getChildContext` function will be called when the state or props changes. In order to update data in the context, trigger a local state update with `this.setState`. This will trigger a new context and changes will be received by the children.
+**fancy-button.js**
+`embed:context/forwarding-refs-fancy-button.js`
 
-```javascript
-import PropTypes from 'prop-types';
+**app.js**
+`embed:context/forwarding-refs-app.js`
 
-class MediaQuery extends React.Component {
-  constructor(props) {
-    super(props);
-    this.state = {type:'desktop'};
-  }
+## Caveats
```

Comment:
Before Caveats, I think we should also add a section that explains how to create a higher order component for consuming context. This is a common sentiment and I think we should proactively address it: https://twitter.com/DankoKozar/status/977469702204162048?s=20

Reply:
I'll add this.

Reply:
Context comes before the Higher-Order Components reference page, but I'll cross link.

Reply:
Yeah, cyclical dependencies are pretty much unavoidable in the "Advanced" section